### PR TITLE
proton-ge-bin: init at 9-1, nixos/steam: use steamcompattool output for extraCompatPackages

### PR DESCRIPTION
**Pull request by Atemu**
## Description of changes

Alternative to https://github.com/NixOS/nixpkgs/pull/295785

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/pr/proton-ge-bin/package.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    ln -s $src $steamcompattool
+
+    runHook postBuild
+  '';
+
+  /*
+    We use the created releases, and not the tags, for the update script as nix-update loads releases.atom
+    that contains both. Sometimes upstream pushes the tags but the Github releases don't get created due to
+    CI errors. Last time this happened was on 8-32, where a tag was created but no releases were created.
```
Comment (Shawn8901) [inline pkgs/by-name/pr/proton-ge-bin/package.nix:31]:
```suggestion
    CI errors. Last time this happened was on 8-33, where a tag was created but no releases were created.
```

actually it was 8-33 https://github.com/GloriousEggroll/proton-ge-custom/releases/tag/GE-Proton8-33  which has no release https://github.com/GloriousEggroll/proton-ge-custom/releases 
i did state that out wrong on the first comment, but when noting it down here, we should be correct with the numbers i guess

File: pkgs/by-name/pr/proton-ge-bin/package.nix [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  outputs = [ "out" "steamcompattool" ];
+
+  buildCommand = ''
+    runHook preBuild
+
+    # Provide an error message when adding it to an env
+    echo "${finalAttrs.pname} should not be installed into environments. Please use programs.steam.extraCompatPackages." > $out
+
+    ln -s $src $steamcompattool
```
Comment (Shawn8901) [inline pkgs/by-name/pr/proton-ge-bin/package.nix:23]: just for me to understand what we are doing here.
Basically we are using fetchzip, that automatically extracts to $src, on  building the derivation we link $src to $steamcompattool.

A small question about the log, shouldnt that also be printed when building the derivation (overall) derivation in hydra later? I mean likely no one might care about that (just to understand it).

File: pkgs/games/steam/fhsenv.nix [hunk 3]
```diff
@@ -309,6 +309,7 @@ in buildFHSEnv rec {
       description = "Steam dependencies (dummy package, do not use)";
     };
 
+  passthru.steamargs = args;
```
Comment (Shawn8901) [inline pkgs/games/steam/fhsenv.nix:312]: that is unrelated to the actual update, isnt it?

File: pkgs/by-name/pr/proton-ge-bin/package.nix (hunk 1 above)
Comment (Atemu) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:31]: This commit is from @NotAShelf's PR; GH just shows it here because it's not in master.

File: pkgs/by-name/pr/proton-ge-bin/package.nix (hunk 2 above)
Comment (Atemu) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:23]: > Basically we are using fetchzip, that automatically extracts to $src, on building the derivation we link $src to $steamcompattool.

Yes, in this case, that's the easiest method. Alternatively, we could copy everything over but I don't see the point in that.

Comment (Atemu) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:23]: > A small question about the log, shouldnt that also be printed when building the derivation (overall) derivation in hydra later? I mean likely no one might care about that (just to understand it).

Which log are you talking about?

Comment (Atemu) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:23]: > and as second question (related to [#295785 (comment)](https://github.com/NixOS/nixpkgs/pull/295785#discussion_r1524227440)):
> 
> > We could symlink that to $out/share/steam/compatibilitytools.d/$name/ too now that I think about it.
> 
> I assume that you dropped that idea, correct? (fine for me, just that i did not overlook something)

We could but I don't really see the point for NixOS purposes tbh.

We'd add `/run/current-system/sw/share/compatibilitytools.d/` to the extra compat paths by default and users would be able to add compat tools to the system-wide enironment.

That'd mostly be there fore users who cling to FHS-isms such as installing software that only one component actually needs into a global path. Making the global env fail to build when doing that simply makes it impossible to be in such a situation; forcing users to utilise the appropriate option.

The only unknown here is non-NixOS purposes where I could see that being useful. I'd say we cross that bridge when we get there and an actual user of that concept needs it.

File: pkgs/games/steam/fhsenv.nix (hunk 3 above)
Comment (Atemu) [reply pkgs/games/steam/fhsenv.nix:312]: Yup, only here so I could easily inspect `extraEnv`.

(To be honest, I didn't actually test this yet at all; all I've done is eval this on the mac and do a few remote builds ;) )

File: pkgs/by-name/pr/proton-ge-bin/package.nix (hunk 2 above)
Comment (Atemu) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:23]: > Yes, in this case, that's the easiest method. Alternatively, we could copy everything over but I don't see the point in that.

One minor cosmetic issue this causes is that it's not obvious which version of proton is being ran when looking at `ps` and the like as it'd just say `/nix/store/0bcbzzxskaw13d378fv3630cv5k09kxf-source/proton waitforexitandrun ...`.

File: pkgs/games/steam/fhsenv.nix (hunk 3 above)
Comment (Atemu) [reply pkgs/games/steam/fhsenv.nix:312]: Did test it now :)

File: pkgs/by-name/pr/proton-ge-bin/package.nix (hunk 1 above)
Comment (Atemu) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:32]: @NotAShelf would you prefer to adjust this in your PR and have it merged on its own or do we want to continue the proton-ge init here?

File: pkgs/by-name/pr/proton-ge-bin/package.nix (hunk 2 above)
Comment (Shawn8901) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:24]: > Which log are you talking about?

I assume that this should be shown in the build logs, isn't it? I mean buildCommand is run when building the derivation eG on hydra or?

Comment (Atemu) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:24]: > I assume that this should be shown in the build logs, isn't it? I mean buildCommand is run when building the derivation eG on hydra or?

What is "this"?

File: nixos/modules/programs/steam.nix [hunk 4]
```diff
@@ -44,8 +44,8 @@ in {
       '';
       apply = steam: steam.override (prev: {
         extraEnv = (lib.optionalAttrs (cfg.extraCompatPackages != [ ]) {
-            STEAM_EXTRA_COMPAT_TOOLS_PATHS = makeBinPath cfg.extraCompatPackages;
-          }) // (prev.extraEnv or {});
+          STEAM_EXTRA_COMPAT_TOOLS_PATHS = makeSearchPathOutput "steamcompattool" "" cfg.extraCompatPackages;
```
Comment (NotAShelf) [inline nixos/modules/programs/steam.nix:47]: why don't we also add bin here?

File: pkgs/by-name/pr/proton-ge-bin/package.nix (hunk 1 above)
Comment (NotAShelf) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:32]: lets continue it here

File: nixos/modules/programs/steam.nix (hunk 4 above)
Comment (Atemu) [reply nixos/modules/programs/steam.nix:47]: Non-executable files such as compat tool description files (vdf) do not belong in `bin/`.

Comment (NotAShelf) [reply nixos/modules/programs/steam.nix:47]: So we're planning to make sure every executable compat tool (i.e stl) provide an additional output?

File: pkgs/by-name/pr/proton-ge-bin/package.nix (hunk 2 above)
Comment (Shawn8901) [reply pkgs/by-name/pr/proton-ge-bin/package.nix:24]: Ah nvm, the line was to long and it did not fit on my screen... so i did not see the ` > $out ` on the echo.
I thought that it would be a log when building, but with the pipe it make more sense to me.
Sorry for the noise.

File: pkgs/games/steam/fhsenv.nix (hunk 3 above)
Comment (Shawn8901) [reply pkgs/games/steam/fhsenv.nix:312]: Can confirm also from my end, steam picks it up correctly and can be used.

File: pkgs/by-name/pr/proton-ge-bin/package.nix [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    As of 2024-03-13, there have been no announcements indicating that the CI has been fixed, and thus
+    we avoid nix-update-script and use our own update script instead.
+    See: <https://github.com/NixOS/nixpkgs/pull/294532#issuecomment-1987359650>
+  */
+  passthru.updateScript = writeScript "update-proton-ge" ''
+    #!/usr/bin/env nix-shell
+    #!nix-shell -i bash -p curl jq common-updater-scripts
+    repo="https://api.github.com/repos/GloriousEggroll/proton-ge-custom/releases"
+    version="$(curl -sL "$repo" | jq 'map(select(.prerelease == false)) | .[0].tag_name' --raw-output)"
+    update-source-version proton-ge-custom "$version"
```
Comment (Shawn8901) [inline pkgs/by-name/pr/proton-ge-bin/package.nix:42]:
```suggestion
    update-source-version proton-ge-bin "$version"
```

that has to reflect the packages attributes name, so i assume currently updating would throw an error

File: nixos/modules/programs/steam.nix (hunk 4 above)
Comment (Atemu) [reply nixos/modules/programs/steam.nix:47]: Yup; an output that only concerns the compat tool files. It wouldn't really be "additional" in most cases though as most compat tools one and only purpose is to act as a compat tool and would therefore have an empty `out` just like what I did to proton-ge-bin.

**Review by NotAShelf — APPROVED** [review]

**Review by Shawn8901 — APPROVED** [review]

**Comment by Atemu** [comment]
Thank you both! :)

**Comment by jh-devv** [comment]
As I said in the older issue:

> As soon as this is merged, I can also add the stuff to build from source.
> Though, I really don't know if this is needed, this would be only for custom upstream WINE patches until VALVe pushes them.

**Comment by NotAShelf** [comment]
Go for it. I've taken a look at building from source this morning, but the build tooling is tedious since they rely *heavily* on docker and build/configure scripts.

**Comment by jh-devv** [comment]
> Go for it. I've taken a look at building from source this morning, but the build tooling is tedious since they rely *heavily* on docker and build/configure scripts.

Yep, I'll look into it sometime next week since it's "not necessary" for the time being

**Comment by andy-bell101** [comment]
Is there some trick to getting this to work? I've included `proton-ge-bin` in the extra steam compat packages field but it doesn't actually show up in steam itself.

**Comment by Shawn8901** [comment]
Did you restart steam afterwards? the module does add an environment variable to the wrapper, that is picked up by steam afterwards.
You might want to check the eviron of a steam pid in /proc to check if those are set correctly.

Otherwise you could check the steam-wrapper.sh it should contain 
```
STEAM_EXTRA_COMPAT_TOOLS_PATHS='/nix/store/sfksnvird46l09bxs2b0inf3c6hkicjw-proton-ge-bin-GE-Proton9-1-steamcompattool/'
```
right before  `exec steam "$@"`

**Comment by andy-bell101** [comment]
Yeah those lines are missing from steam-wrapper.sh

**Comment by Atemu** [comment]
The changes are in nixos-unstable and you'd get an error if you tried to set this without the change. Make sure you're using the programs.steam setting an not getting steam from elsewhere (systemPackages or home-manager).

**Comment by andy-bell101** [comment]
Ah yeah I was picking up steam from system packages. Removing that fixed it. Thanks for the advice

**Comment by Petrusion** [comment]
> As I said in the older issue:
> 
> > As soon as this is merged, I can also add the stuff to build from source.
> > Though, I really don't know if this is needed, this would be only for custom upstream WINE patches until VALVe pushes them.

@sofiedotcafe 
Just wanted to ask if you'd still be open to adding the option to build this from source. I’d find it quite useful, I like to override performance sensitive packages with additional compile flags (like cpu arch), and being able to add custom WINE patches would also be very helpful. 

Of course, only if you have the time and interest.

**Comment by Atemu** [comment]
Please open a packaging request to discuss this.
